README.md
# php-http

<a href="https://github.com/josephscott/php-http/actions"><img src="https://github.com/josephscott/php-http/actions/workflows/tests.yml/badge.svg"></a>

For those times when you need a simple set of HTTP request functions, with no extras needed.  It is build upon `file_get_contents()`, keeping the easy URL requests easy.

## Install

`composer require josephscott/php-http`

Alternatively you can grab a copy of `https://raw.githubusercontent.com/josephscott/php-http/trunk/src/http.php` and `require` it as needed.

## Testing
`make test`

## Usage

```php
require __DIR__ . '/vendoer/autoload.php';

// Object style GET
$http = new HTTP();
$response = $http->get( url: 'https://httpbin.org/get' );
if ( $response['error'] ) {
	// Handle request error
}

// Object style POST
$response = $http->post(
	url: 'https://httpbin.org/post',
	headers: [ 'x-custom' => 'thing' ],
	data: [ 'hello' => 'world' ],
);

// Static style GET
$response = HTTP::get( url: 'https://httpbin.org/get' );

// Static style Post
$response = HTTP::post( url: 'https://httpbin.org/post' );
```

### The response array

The response is always an array, with the following fields:

- `error`: boolean
- `headers`: array of name/value pairs for the response headers
- `body`: string of the response body
- `total_time`: milliseconds it took for the entire request to complete


CHANGELOG.md
# Changelog

## 0.1.1 - 2023-04-21
- Work with composer via classmap

## 0.1.0 - 2023-04-20
- Major changes to the API, reducing the amount of code and making the API more consistent across the static and object methods
- Static tests have been merged into a single set per verb
- Readme examples have been updated to use the new API
- Simplify body encoding checks
- Makefile test failure cleanup improved
- Add `Connection: close` to the default headers to avoid keep-alive delays
- Limit HTTP calls to the list of known methods
- Reduce the `get: slow` sleep time, reducing the amount of time the test run takes
- New tests for delete() method

## 0.0.6 - 2023-04-18
- Support for PUT requests via put() method
- New test for put() method
- Support for PUT requests via pu() method
- New test for pu() method

## 0.0.5 - 2023-04-14
- Notes on how to use the HTTP object and class
- Provide return types
- New test for changing the default user agent
- Support for HEAD requests via head() method
- New test for head() method
- Support for HEAD requests via h() method
- New test for h() method
- New test for get with a URL that has no web server running

## 0.0.4 - 2023-04-12
- Github Actions for tests
- Static method, HTTP::g() for GET
- New test for HTTP::g()
- Static method, HTTP::p() for POST
- New test for HTTP::p()
- New simple PHP server that the tests run against, all tests have been updated to use it

## 0.0.3 - 2023-04-10
- Provide default HTTP options
- Default to HTTP/1.1
- Run PHP lint as part of the tests
- Use the `ignore_errors` option instead of the error handler to capture response error conditions
- New test: `get: fail, 500`
- Add timeout support, defaults to 90 seconds
- New test: `get: timeout`
- New public method: request().  This centralizes the file_get_contents() call into one place.
- Update the get() and post() methods to use the new request() method
- Track the total time a request took
- New test: `get: slow`, make sure total time tracking works

## 0.0.2 - 2023-04-08
- Rename to HTTP
- Lowercase all response header names
- Set types for function args
- Add support for custom request headers
- Move everything to a single class
- Add private build_context() method
- List of default request headers, currently user_agent and accept
- Include the HTTP response code in the list of response headers
- Testing with phpstan
- Testing with pest
- Handle PHP warnings with failed HTTP requests

## 0.0.1 - 2023-04-06
- Basic minimal working version


tests/Pest.php
<?php
declare( strict_types = 1 );

require __DIR__ . '/../src/http.php';


tests/GTest.php
<?php
declare( strict_types = 1 );

test( 'HTTP::get', function() {
	$response = HTTP::get( url: 'http://127.0.0.1:7878/?method=get' );

	expect( $response['error'] )->toBe( false );
	expect( $response['headers']['response_code'] )->toBe( 200 );
	expect( $response['headers']['content-type'] )->toBe( 'application/json' );
} );


tests/HTest.php
<?php
declare( strict_types = 1 );

test( 'HTTP::head', function() {
	$response = HTTP::head( url: 'http://127.0.0.1:7878/?method=head' );

	expect( $response['error'] )->toBe( false );
	expect( $response['headers']['response_code'] )->toBe( 200 );
	expect( $response['headers']['content-type'] )->toBe( 'application/json' );
} );


tegstop.php
<?php
class Tegstop {
	public function __construct() { }

	public function get( string $url, array $headers = [] ) {
		$response = [];

		$options = [
			'http' => [
				'method' => 'GET',
				'header' => ''
			]
		];
		foreach ( $headers as $header_name => $header_value ) {
			$options['http']['header'] .= "$header_name: $header_value\r\n";
		}

		$context = stream_context_create( $options );
		$body = file_get_contents( $url, false, $context );

		$response['error'] = false;
		$response['headers'] = self::parse_headers( $http_response_header );
		$response['body'] = $body;

		if ( $body === false ) {
			$response['error'] = true;
			return $response;
		}

		return $response;
	}

	public function post( string $url, array $headers = [], array $data = [] ) {
		$response = [];

		$options = [
			'http' => [
				'method' => 'POST',
				'header' => '',
				'content' => http_build_query( $data ),
			],
		];
		$options['http']['header'] .= 'Content-Type: '
			. "application/x-www-form-urlencoded\r\n";
		foreach ( $headers as $header_name => $header_value ) {
			$options['http']['header'] .= "$header_name: $header_value\r\n";
		}

		$context = stream_context_create( $options );
		$body = file_get_contents( $url, false, $context );

		$response['error'] = false;
		$response['headers'] = self::parse_headers( $http_response_header );
		$response['body'] = $body;

		if ( $body === false ) {
			$response['error'] = true;
			return $response;
		}

		return $response;
	}

	/* Private */

	private function parse_headers( array $headers ) {
		$parsed = [];

		foreach ( $headers as $header ) {
			$parts = explode( ':', $header, 2 );
			if ( count( $parts ) === 2 ) {
				$parsed[ strtolower( trim( $parts[0] ) ) ] = trim( $parts[1] );
			}
		}

		return $parsed;
	}
}


src/http.php
<?php
declare( strict_types = 1 );

class HTTP {
	// https://www.php.net/manual/en/context.http.php
	public $default_options = [
		'protocol_version' => 1.1,
		'method' => 'GET',
		'header' => '',
		'ignore_errors' => true,
		'timeout' => 90,
	];

	public $default_headers = [
		'user_agent' => 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/112.0.0.0 Safari/537.36 php-http/0.1.1',
		'accept' => '*/*',
		'connection' => 'close'
	];

	public function __construct() { }

	public function __call(
		string $name,
		array $args
	):array {
		$response = $this->make_request(
			name: $name,
			args: $args
		);
		return $response;
	}

	public static function __callStatic( string $name, array $args ) {
		$http = new HTTP();
		$response = $http->make_request(
			name: $name,
			args: $args
		);
		return $response;
	}

	public function make_request( string $name, array $args ) {
		$url = $args['url'] ?? $args[0];
		$headers = $args['headers'] ?? $args[1] ?? [];
		$data = $args['data'] ?? $args[2] ?? [];

		$response = $this->request(
			method: strtoupper( $name ),
			url: $url,
			headers: $headers,
			data: $data
		);

		return $response;
	}

	public function request(
		string $method,
		string $url,
		array $headers = [],
		array $data = []
	):array {
		$response = [
			'error' => false,
			'headers' => [],
			'body' => '',
		];

		// https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods
		$valid_methods = [
			'GET',
			'HEAD',
			'POST',
			'PUT',
			'DELETE',
			'CONNECT',
			'OPTIONS',
			'TRACE',
			'PATCH'
		];
		if ( !in_array( $method, $valid_methods, true ) ) {
			return $response;
		}

		$context = $this->build_context( method: $method, body: $data );

		// XXX: HACK
		// Make Pest happy by suppressing the warnings that can happen
		// I'd like to find a way to deal with warnings without using @
		$start = microtime( true );
		$body = @file_get_contents(
			filename: $url,
			use_include_path: false,
			context: $context
		);
		$response['total_time'] = number_format(
			( microtime( true ) - $start ),
			6
		);
		if ( $body === false ) {
			$response['error'] = true;
			return $response;
		}

		$response['headers'] = self::parse_headers( $http_response_header );
		$response['body'] = $body;

		if (
			$response['headers']['response_code'] < 200
			|| $response['headers']['response_code'] > 299
		) {
			$response['error'] = true;
			return $response;
		}

		return $response;
	}

	/* Private */

	private function build_context(
		string $method,
		array $headers = [],
		array $body = []
	) {
		$options = [ 'http' => $this->default_options ];
		$options['http']['method'] = $method;

		$headers = array_merge( $this->default_headers, $headers );

		if ( !empty( $body ) ) {
			$options['http']['content'] = http_build_query( $body );
			$headers['Content-Type'] = 'application/x-www-form-urlencoded';
		}

		foreach ( $headers as $header_name => $header_value ) {
			$options['http']['header'] .= "$header_name: $header_value\r\n";
		}

		$context = stream_context_create( $options );
		return $context;
	}

	private function parse_headers( array $headers ):array {
		$parsed = [];

		$response_code = array_shift( $headers );
		if ( preg_match( '#HTTP/[0-9\.]+\s+([0-9]+)#', $response_code, $matches ) ) {
			$headers[] = 'response_code: ' . intval( $matches[1] );
		}

		foreach ( $headers as $header ) {
			$parts = explode( ':', $header, 2 );
			if ( count( $parts ) === 2 ) {
				$parts[1] = trim( $parts[1] );
				if ( is_numeric( $parts[1] ) ) {
					$parts[1] = (int) $parts[1];
				}

				$parsed[ strtolower( trim( $parts[0] ) ) ] = $parts[1];
			}
		}

		return $parsed;
	}
}
